src/main/java/org/networkcalculus/dnc/AnalysisConfig.java

```diff
@@ -36,188 +37,212 @@
+	}
+
+	public void clearArrivalBoundMethods() {
+		arrival_bound_methods.clear();
+	}
+
+	public void setArrivalBoundMethod(ArrivalBoundMethod arrival_bound_method) {
+		clearArrivalBoundMethods();
+		arrival_bound_methods.add(arrival_bound_method);
+	}
+
+	public Set<ArrivalBoundMethod> arrivalBoundMethods() {
+		return new HashSet<ArrivalBoundMethod>(arrival_bound_methods);
+	}
+
+	public void setArrivalBoundMethods(Set<ArrivalBoundMethod> arrival_bound_methods_set) {
+		clearArrivalBoundMethods();
+		arrival_bound_methods.addAll(arrival_bound_methods_set);
+	}
+
+	public void addArrivalBoundMethod(ArrivalBoundMethod arrival_bound_method) {
+		arrival_bound_methods.add(arrival_bound_method);
+	}
+
+	public void addArrivalBoundMethods(Set<ArrivalBoundMethod> arrival_bound_methods_set) {
+		arrival_bound_methods.addAll(arrival_bound_methods_set);
+	}
+
+	public boolean removeArrivalBoundMethod(ArrivalBoundMethod arrival_bound_method) {
+		if (arrival_bound_methods.size() == 1) {
+			return false;
+		} else {
+			return arrival_bound_methods.remove(arrival_bound_method);
+		}
+	}
+
+	public boolean convolveAlternativeArrivalBounds() {
+		return convolve_alternative_arrival_bounds;
+	}
+
+	public void setConvolveAlternativeArrivalBounds(boolean convolve_alt_abs) {
+		convolve_alternative_arrival_bounds = convolve_alt_abs;
+	}
+
+	public boolean serverBacklogArrivalBound() {
+		return server_backlog_arrival_bound;
+	}
+
+	public void setServerBacklogArrivalBound(boolean server_backlog_arrival_bound) {
+		this.server_backlog_arrival_bound = server_backlog_arrival_bound;
+	}
+
+	public boolean useFlowProlongation() {
+		return flow_prolongation;
+	}
+
+	public void setUseFlowProlongation(boolean prolong_flows) {
+		flow_prolongation = prolong_flows;
+	}
+
+	/**
+	 * Returns a deep copy of this analysis configuration.
+	 *
+	 * @return The copy.
+	 */
+	public AnalysisConfig copy() { // deep copy as primitive data types are copied by value
+		return new AnalysisConfig(multiplexing_enforcement, enforce_max_sc, enforce_max_sc_output_rate, arrival_bound_methods,
+				convolve_alternative_arrival_bounds, server_backlog_arrival_bound);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(multiplexing_enforcement, enforce_max_sc, enforce_max_sc_output_rate, arrival_bound_methods,
+				convolve_alternative_arrival_bounds, server_backlog_arrival_bound);
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null)
+			return false;
+		if (getClass() != obj.getClass())
+			return false;
+		AnalysisConfig other = (AnalysisConfig) obj;
+
+		return Objects.equals(this.multiplexing_enforcement, other.multiplexing_enforcement) &&
+				Objects.equals(this.enforce_max_sc, other.enforce_max_sc) &&
+				Objects.equals(this.enforce_max_sc_output_rate, other.enforce_max_sc_output_rate) &&
+				Objects.equals(this.arrival_bound_methods, other.arrival_bound_methods) &&
```

Comment:
I am not sure the object.equals works as we expect it on this set: I suspect that it will check if we have the same set instantiation but we assume equality of these two conditions are met:
1) The sets have the same amount of items
2) One set contains all the items of the other set

The set items are enum entries, i.e., primitive types that can again be compared with equals() and thus the 2) should work fine.

Reply:
**This code is compliant with these two requirements that you have described!**

I'm using **Java JDK 12** sources as reference for checking Java's code behavior:

The `HashSet<E>` class does not have `equals()` overridden in there, but it is overridden in `AbstractSet<E>` which is its 
super class.

This is part of `HashSet<E>` class implementation:
```
public class HashSet<E> extends AbstractSet<E> implements Set<E>, Cloneable, java.io.Serializable {
...
}
```

This is the `equals()` implementation in `AbstractSet<E>`:

```
    /**
     * Compares the specified object with this set for equality.  Returns
     * {@code true} if the given object is also a set, the two sets have
     * the same size, and every member of the given set is contained in
     * this set.  This ensures that the {@code equals} method works
     * properly across different implementations of the {@code Set}
     * interface.<p>
     *
     * This implementation first checks if the specified object is this
     * set; if so it returns {@code true}.  Then, it checks if the
     * specified object is a set whose size is identical to the size of
     * this set; if not, it returns false.  If so, it returns
     * {@code containsAll((Collection) o)}.
     *
     * @param o object to be compared for equality with this set
     * @return {@code true} if the specified object is equal to this set
     */
    public boolean equals(Object o) {
        if (o == this)
            return true;

        if (!(o instanceof Set))
            return false;
        Collection<?> c = (Collection<?>) o;
        if (c.size() != size())
            return false;
        try {
            return containsAll(c);
        } catch (ClassCastException | NullPointerException unused) {
            return false;
        }
    }
```

Reply:
Great, I didn't know (and didn't check either). Thanks for pointing it out.